### ISSUE TITLE
Default scene session data is mutatable within library

### ISSUE THREAD

**Issue by wuiyang** [telegraf/telegraf#1493]
## Context

* Telegraf.js Version: 4.4.0
* Node.js Version: v12.22.1
* Operating System: Windows 10 20H2 build 19042.1110

## Minimal Example Code Reproducing the Issue

```ts
  const greeterScene = new Scenes.BaseScene<Scenes.SceneContext>('greeter');
  greeterScene.enter(async ctx => {
    await ctx.reply(`Hi, I'm in ${ctx.session.__scenes.current} scene`);
  });
  greeterScene.on('text', async ctx => {
    await ctx.reply(`I'm still in ${ctx.session.__scenes.current} scene`);
    await ctx.scene.enter('echo');
    await ctx.reply('After enter echo');
  });
  greeterScene.leave(async ctx => {
    await ctx.reply(`Bye, I'm in ${ctx.session.__scenes.current} scene`);
  });

  const echoScene = new Scenes.BaseScene<Scenes.SceneContext>('echo');
  echoScene.enter(async ctx => {
    await ctx.reply(`Echo, I'm in ${ctx.session.__scenes.current} scene`);
    await ctx.scene.leave();
    await ctx.reply('After leave echo');
  });
  echoScene.on('text', async ctx => {
    await ctx.reply(`Echo, I'm still in ${ctx.session.__scenes.current} scene`);
  });
  echoScene.leave(async ctx => {
    await ctx.reply(`Echo, Bye, I'm in ${ctx.session.__scenes.current} scene`);
  });

  const stage = new Scenes.Stage<Scenes.SceneContext>([
    greeterScene,
    echoScene,
  ]);

  bot.use(stage.middleware() as never);
  bot.command('/greet', ctx =>
    (ctx as never as Scenes.SceneContext).scene.enter('greeter'),
  );
  bot.command('/echo', ctx =>
    (ctx as never as Scenes.SceneContext).scene.enter('echo'),
  );
  bot.on('text', ctx => ctx.reply("I'm not in scene"));
```

Setup bot and setup command + scene with code above.
Reply `/greet`, then reply any text twice.

## Expected Behavior

The bot should enter greeter scene when received `/greet`.
Then the bot should enter echo scene and leave after completed when received any text messages in greeter scene.
Then the bot should fall back to default for all message after.
The message received should be:
```
me: /greet
bot: Hi, I'm in greeter scene

me: d
bot: I'm still in greeter scene
bot: Bye, I'm in greeter scene
bot: Echo, I'm in echo scene
bot: Echo, Bye, I'm in echo scene
bot: After leave echo
bot: After enter echo

me: d
bot: I'm not in scene
```

## Current Behavior

Reply any text message after `/greet` will stuck in echo module. Receiving 'Echo, I'm still in echo scene' for every message replied.
The message received:
```
me: /greet
bot: Hi, I'm in greeter scene

me: d
bot: I'm still in greeter scene
bot: Bye, I'm in greeter scene
bot: Echo, I'm in echo scene
bot: Echo, Bye, I'm in echo scene
bot: After leave echo
bot: After enter echo

me: d
bot: Echo, I'm still in echo scene
```

## Error Message and Logs (`export DEBUG='telegraf:*'`)
Unable to get any internal log. However, I have found that the [scene context handler](https://github.com/telegraf/telegraf/blob/develop/src/scenes/context.ts#L82) might have some hidden error, where it replaces `session` with `this.options.defaultSession` directly, which is prune to mutating default value from external side, which line 97 could mutate it internally.

```
Please include any relevant log snippets or files here.
```
Screenshot of `this.options.defaultSession` value after leaving scene in echo scene.
![image](https://user-images.githubusercontent.com/2557383/127748957-4aabb86b-cc58-4ac3-9460-0c8033220bdf.png)

Screenshot of mutating `this.session` will mutate `this.options.defaultSession`
![image](https://user-images.githubusercontent.com/2557383/127749112-834bc6dc-2178-4ec0-b1b3-cbfd3e149c8e.png)

**Comment by NormandoHall** [comment]
Same here. I was fixed cleaning session after awaiting leave:

```js
await ctx.scene.leave();
ctx.session = {};

```

**Comment by NormandoHall** [comment]
Try this. I was found that entering scenes with async, completes the update process. 

```ts
  bot.command('/greet', async ctx =>
    await (ctx as never as Scenes.SceneContext).scene.enter('greeter'),
  );
  bot.command('/echo', async ctx =>
    await (ctx as never as Scenes.SceneContext).scene.enter('echo'),
  );
```

**Comment by MKRhere** [comment]
Fixed in #1714
